### ISSUE TITLE
Range attributes of predicates still hitting in queries after being modified

### ISSUE THREAD

**Issue by adriabonetmrf** [vespa-engine/vespa#18637]
**Describe the bug**
There is a weird behaviour whenever using range attributes in predicates, seems like when we modify with either a **partial update or full update** the predicate on a document the old range value remains in the index and the document is returned when querying by that value or the new one.

For a predicate field in the schema with the following specification:
````
field show_condition type predicate {
        indexing: summary | attribute
        index {
            arity: 2  # mandatory
        }
}
````

- Given a document with field `show_condition` = `'value' in [1000..1000]`
- Querying by `predicate({},{"value": 1000L})` returns the document
- Then changing `show_condition` to `'value' in [2000..2000]`
- Querying by `predicate({},{"value": 2000L})`  returns the document and querying by `predicate({},{"value": 1000L})` still returns the document

This doesn’t happen with regular attributes so it seems a problem with the index not cleaning up old range entries

**To Reproduce**
Steps to reproduce the behaviour:
1. Create a document of an schema containing the predicate field above with a predicate targeting a range attribute on a fixed value ([1..1])
2. Query the document `select * from documents where predicate({},{"value": 1L})` and check its hit
3. Modify the document and change its predicate to another value ([2..2])
4. Query again `select * from documents where predicate({},{"value": 1L})` and see that it's still getting hit

**Expected behavior**
The document shouldn't get hit on step 4 as the targeted range has been changed in the predicate.

**Environment (please complete the following information):**
OS: Ubuntu 20.04.1 with docker 20.10.3
Infrastructure: Self-hosted

**Vespa version**
vespaengine/vespa:7.419.22

**Comment by jobergum** [comment]
In step 3 what is the operation? Update assign?

**Comment by adriabonetmrf** [comment]
Yes, but the same behaviour is present whenever I perform a whole document write, so it can be either a put or an update.

**Comment by bjormel** [comment]
Thanks for reporting this issue, we are able to reproduce this.

We will triage this issue next week, or the week after.

**Comment by baldersheim** [comment]
@adriabonetmrf The fix(#18212) for bug #17481 was included until vespa-7.419.32. I see that you used 7.419.22 which does not have the fix.
Are you able to reproduce with a newer vespa version ?

**Comment by baldersheim** [comment]
I have verified in system test that there is indeed an issue with range attributes in predicates. I will continue and reproduce in unit test to identify root cause.

**Comment by baldersheim** [comment]
This seems to only be an issue with single point ranges.
Multipoint ranges test added in https://github.com/vespa-engine/system-test/pull/1835

**Comment by adriabonetmrf** [comment]
Thanks for the update @baldersheim, I'll check the version again as I thought we were on the right one after upgrading a couple of days after your release, also we didn't detect the bug coming back until some days ago with a low amount of documents impacted so this might explain that behaviour.

**Comment by baldersheim** [comment]
@adriabonetmrf I have verified that this bug was not fixed by the first fix. This is still an issue. I am working on narrowing down the scenarios. But from what I can reproduce it is related to ranges with a single point. Fx [7...7]. If you at one point have a document with a single point interval this point is still visible after the document has been updated to either a single point interval like [8..8] or [9...20]. 7 will still yield a hit.

**Comment by adriabonetmrf** [comment]
@baldersheim yes, exactly that's the one I opened this issue for. I was mentioning the old one in the last comment tho as you pointed out we are not in the correct release, still it's weird, we updated with the release you made on 15th june (https://github.com/vespa-engine/vespa/releases/tag/vespa-7.419.22-1) which is 4 days after your comment giving the version in which the fix was applied, can you doublecheck to be sure there wasn't a typo at some point and link me to the correct version with the fix applied so we can update? Thanks

**Comment by baldersheim** [comment]
The fix was merged on 7.419.32. But if the release of 7.419 was the .22 it did not contain the fix. Then you nede to get a version 7.420 or later. Just take the latest version available. Currently it seems to be https://github.com/vespa-engine/vespa/releases/tag/vespa-7.447.10-1

**Comment by adriabonetmrf** [comment]
Ok, thanks @baldersheim, I guess there was a misunderstanding on our part when updating.

**Comment by baldersheim** [comment]
Root cause fixed in https://github.com/vespa-engine/vespa/pull/18714
The fix will be part of 7.450.8. So the first release after that should be fine.

**Comment by baldersheim** [comment]
Verified that both unit tests and system tests pass with 7.450.8.
